[libc][math][c23] Implement canonicalize functions

Sh0g0-1758 opened the pull request:
Fixes: #85286

Sh0g0-1758 committed:
Initial commit

Sh0g0-1758 committed:
Initial implementation

Sh0g0-1758 committed:
Added space

Sh0g0-1758 committed:
Ran formatter

Sh0g0-1758 committed:
Formatter

Sh0g0-1758 committed:
minor updates

Sh0g0-1758 committed:
Cleaned the code

Sh0g0-1758 commented:
@lntue, creating this to confirm whether or not this is moving in the right direction. I am little confused in the implementation of the long double type. I will add tests once you confirm about the implementation of others. As you said  :

> float, double, and float128: these are IEEE 754 standard formats, most of the inputs are already canonical, with signaling NaNs are the only exception.

So I figured that they are already in canonical format and so I just need to assign *x to *cx.

lntue commented on libc/config/linux/x86_64/entrypoints.txt:
```diff
@@ -342,6 +342,10 @@ set(TARGET_LIBM_ENTRYPOINTS
     libc.src.math.atanhf
+    libc.src.math.canonicalize
+    libc.src.math.canonicalizef
+    libc.src.math.canonicalizef128
```
float128 is guarded, and their entry points are added in lines 480+ below.

lntue commented on libc/src/math/canonicalize.h:
```diff
@@ -0,0 +1,19 @@
+//===-- Implementation header for canonicalize ------------------------*- C++
+//-*-===//
```
format

Sh0g0-1758 replied:
done.

lntue commented on libc/spec/stdc.td:
```diff
@@ -570,6 +570,11 @@ def StdC : StandardSpec<"stdc"> {
+
+          FunctionSpec<"canonicalize", RetValSpec<IntType>, [ArgSpec<DoubleType>, ArgSpec<DoubleType>]>,
+          FunctionSpec<"canonicalizef", RetValSpec<IntType>, [ArgSpec<FloatType>, ArgSpec<FloatType>]>,
+          FunctionSpec<"canonicalizef128", RetValSpec<IntType>, [ArgSpec<Float128Type>, ArgSpec<Float128Type>]>,
```
f128 variant uses `GuardedFunctionSpec` instead

Sh0g0-1758 replied:
done.

lntue commented on libc/src/math/generic/canonicalizef.cpp:
```diff
@@ -0,0 +1,26 @@
+namespace LIBC_NAMESPACE {
+
+LLVM_LIBC_FUNCTION(int, canonicalizef, (float *cx, const float *x)) {
+  using FPB = fputil::FPBits<float>;
```
These implementations are identical.  You can refactor them to a template in https://github.com/llvm/llvm-project/blob/main/libc/src/__support/FPUtil/BasicOperations.h

lntue commented on libc/src/math/generic/canonicalizel.cpp:
```diff
@@ -0,0 +1,27 @@
+
+LLVM_LIBC_FUNCTION(int, canonicalizel,
+                   (long double *cx, const long double *x)) {
+  using FPB = fputil::FPBits<long double>;
```
You will need a separate implementation for x86-long double, following the description at https://en.wikipedia.org/wiki/Extended_precision#x86_extended_precision_format

Sh0g0-1758 replied:
Following from this code : 

```cpp
  else if constexpr (cpp::is_same_v<UnqualT, long double>) {
    if constexpr (__LDBL_MANT_DIG__ == 53)
      return FPType::IEEE754_Binary64;
    else if constexpr (__LDBL_MANT_DIG__ == 64)
      return FPType::X86_Binary80;
    else if constexpr (__LDBL_MANT_DIG__ == 113)
      return FPType::IEEE754_Binary128;
  }
```

I think I can add the logic for long double in `BasicOperations.h` itself with a simple if a modification to check for x86_80 like so : 

```cpp
template <typename T, cpp::enable_if_t<cpp::is_floating_point_v<T>, int> = 0>
LIBC_INLINE T canonicalize(T *cx, const T *x) {
  if constexpr (get_fp_type<T>() == FPType::X86_Binary80) {
    // x86 extended precision has a 64-bit mantissa.
    
  }
  FPBits<T> sx(*x);
  if (sx.is_signaling_nan()) {
    *cx = FPBits<T>::quiet_nan();
  } else {
    *cx = *x;
  }
  return 0;
}
```

Since the other types for long double, ie. `IEEE754_Binary64` and `IEEE754_Binary128` are already in canonical form. But from my understanding/research, X86_80 is the same as IEEE754_Binary80 rendering it already in canonical form. I read through the wiki page but that was the explanation for X86_80 special implementation given in `FPBits.h`. Can you please explain what specific transformation is required here, if any.

Sh0g0-1758 replied:
I think after this, I can now start writing tests and mark the PR as ready for review.

lntue replied:
From the table **Interpretation of the fields of an x86 Extended Precision value** on the [wiki page](https://en.wikipedia.org/wiki/Extended_precision#x86_extended_precision_format), all the **pseudo** and **unnormal** are not canonical forms, and you will need to convert them to the canonical forms of the same values.

Sh0g0-1758 replied:
Ah, thanks for this. Added an implementation for the same.

jcranmer-intel commented:
Some drive-by comments:

* canonicalize(SNAN) needs to raise the INVALID floating point exception
* it's a should, not a must requirement, but the corresponding quiet NaN should have the same payload as the signaling NaN (i.e., only the bit indicating whether or not the NaN is an sNaN should change)
* x86 and PPC long double types have non-canonical values that need to be handled. I don't know enough about the PPC long double type to give guidance here, but the x86 long double type considers all non-canonical values to be invalid, and will generally refuse to operate on them in the first place--glibc in this case returns 1 and does nothing.

lntue commented:
> Some drive-by comments:
> 
> * canonicalize(SNAN) needs to raise the INVALID floating point exception
> * it's a should, not a must requirement, but the corresponding quiet NaN should have the same payload as the signaling NaN (i.e., only the bit indicating whether or not the NaN is an sNaN should change)
> * x86 and PPC long double types have non-canonical values that need to be handled. I don't know enough about the PPC long double type to give guidance here, but the x86 long double type considers all non-canonical values to be invalid, and will generally refuse to operate on them in the first place--glibc in this case returns 1 and does nothing.

Thanks @jcranmer-intel !  I totally forgot to include error handling section F.10.8.7 (N3096) of canonicalize functions to the issue's description.

About the PPC long double, we will skip it for now and focusing on x86-64, arm, and riscv, which we have the hardware to test directly.  Hopefully, someone with PPC knowledge and hardware could help us to fill it in later.

Sh0g0-1758 committed:
Moved logic to BasicOperations.h

Sh0g0-1758 committed:
formatter

Sh0g0-1758 committed:
Updated EntryPoints.txt

Sh0g0-1758 committed:
Format header

lntue commented on libc/spec/stdc.td:
```diff
@@ -570,6 +570,11 @@ def StdC : StandardSpec<"stdc"> {
+
+          FunctionSpec<"canonicalize", RetValSpec<IntType>, [ArgSpec<DoubleType>, ArgSpec<DoubleType>]>,
+          FunctionSpec<"canonicalizef", RetValSpec<IntType>, [ArgSpec<FloatType>, ArgSpec<FloatType>]>,
+          GuardedFunctionSpec<"canonicalizef128", RetValSpec<IntType>, [ArgSpec<Float128Type>, ArgSpec<Float128Type>]>,
```
Add the guard `LIBC_TYPES_HAS_FLOAT128` at the end, similar to L572.

Sh0g0-1758 replied:
done.

Sh0g0-1758 committed:
Added guard

Sh0g0-1758 committed:
Formatter

lntue commented on libc/src/__support/FPUtil/BasicOperations.h:
```diff
@@ -73,6 +73,17 @@ LIBC_INLINE T fdim(T x, T y) {
+LIBC_INLINE T canonicalize(T *cx, const T *x) {
+  FPBits<T> sx(*x);
+  if (sx.is_signaling_nan()) {
+    *cx = FPBits<T>::quiet_nan();
```
As @jcranmer-intel mentioned, you'll need to copy the NaN payload.  Simply extract the sign and the mantissa of `sx`, then pass them to `FPBits<T>::quiet_nan(sign, mantissa)` should work.  And you'll need to raise `FE_INVALID` exception, using `raise_except_if_required` function in `FEnvImpl.h` (add the needed `#include` and dependency in `CMakeLists.txt` also)

Sh0g0-1758 replied:
done.

Sh0g0-1758 committed:
Error handling

lntue commented on libc/src/__support/FPUtil/BasicOperations.h:
```diff
@@ -73,6 +73,17 @@ LIBC_INLINE T fdim(T x, T y) {
+template <typename T, cpp::enable_if_t<cpp::is_floating_point_v<T>, int> = 0>
+LIBC_INLINE T canonicalize(T *cx, const T *x) {
+  FPBits<T> sx(*x);
+  if (sx.is_signaling_nan()) {
```
We can optimize it a bit by putting this rare condition under `LIBC_UNLIKELY` (from `src/__support/macros/optimization.h` header)

Sh0g0-1758 replied:
done.

Sh0g0-1758 committed:
Formatter

Sh0g0-1758 committed:
Corrected Error handling

Sh0g0-1758 committed:
Added LIBC_UNLIKELY

lntue commented on libc/src/__support/FPUtil/BasicOperations.h:
```diff
@@ -73,6 +75,20 @@ LIBC_INLINE T fdim(T x, T y) {
 }
 
+template <typename T, cpp::enable_if_t<cpp::is_floating_point_v<T>, int> = 0>
+LIBC_INLINE T canonicalize(T *cx, const T *x) {
```
Change the parameter types to `T& c` and `const T &x`.  Update the call sites accordingly.

Sh0g0-1758 replied:
done.

lntue commented on libc/src/math/generic/canonicalize.cpp:
```diff
@@ -0,0 +1,19 @@
+namespace LIBC_NAMESPACE {
+
+LLVM_LIBC_FUNCTION(int, canonicalize, (double *cx, const double *x)) {
+  return fputil::canonicalize(&cx, &x);
```
Once `fputil::canonicalize` parameters become `T& c, const T&x`, these call sites should be:
```
  fputil::canonicalized(*cx, *x);
```

Sh0g0-1758 committed:
Changed references

Sh0g0-1758 committed:
Added Canonical X86_80

Sh0g0-1758 committed:
Added tests

Sh0g0-1758 committed:
Bug fixes

Sh0g0-1758 committed:
Merge branch 'llvm:main' into 85286

lntue commented on libc/src/__support/FPUtil/BasicOperations.h:
```diff
@@ -73,6 +76,54 @@ LIBC_INLINE T fdim(T x, T y) {
+    cx = FPBits<T>::quiet_nan(sx.sign(), sx.get_explicit_mantissa()).get_val();
+    raise_except_if_required(FE_INVALID);
+    return 1;
+  } else if constexpr (get_fp_type<T>() == FPType::X86_Binary80) {
```
since you already have early return, just remove the `else`, making it:
```
if (LIBC_UNLIKELY(sx.is_signaling_nan()) {
}

if constexpr (get_fp_type<T>() != FPType::X86_Binary80) {
  cx = x;
  return 0;
}

// x86-extended precision
...
```

Sh0g0-1758 replied:
done.

lntue commented on libc/src/__support/FPUtil/BasicOperations.h:
```diff
@@ -73,6 +76,54 @@ LIBC_INLINE T fdim(T x, T y) {
+LIBC_INLINE int canonicalize(T &cx, const T &x) {
+  FPBits<T> sx(x);
+  if (LIBC_UNLIKELY(sx.is_signaling_nan())) {
+    cx = FPBits<T>::quiet_nan(sx.sign(), sx.get_explicit_mantissa()).get_val();
```
Does it work correctly for 80-bit extended precision?

Sh0g0-1758 replied:
Changed the order of if-else now.

lntue commented on libc/src/__support/FPUtil/BasicOperations.h:
```diff
@@ -73,6 +76,54 @@ LIBC_INLINE T fdim(T x, T y) {
+        if (!bit61) {
+          cx = FPBits<T>::inf().get_val();
+        } else {
+          cx = FPBits<T>::signaling_nan().get_val();
```
Treat this the same as signaling nans: return `quiet_nan` with the same payload and raise `FE_INVALID`.

Sh0g0-1758 replied:
done.

lntue commented on libc/src/__support/FPUtil/BasicOperations.h:
```diff
@@ -73,6 +76,54 @@ LIBC_INLINE T fdim(T x, T y) {
+          cx = FPBits<T>::signaling_nan().get_val();
+        }
+      } else if (!bit63 && bit62) {
+        cx = FPBits<T>::signaling_nan().get_val();
```
Treat this the same as signaling nans: return `quiet_nan` with the same payload and raise `FE_INVALID`.

lntue commented on libc/src/__support/FPUtil/BasicOperations.h:
```diff
@@ -73,6 +76,54 @@ LIBC_INLINE T fdim(T x, T y) {
+    } else if (exponent == 0 && bit63) {
+      cx = FPBits<T>::get_canonical_val(sx.sign(), mantissa).get_val();
+    } else if (!bit63) {
+      cx = FPBits<T>::get_canonical_val(sx.sign(), mantissa).get_val();
```
Use `FPBits<T>::make_value` instead: https://github.com/llvm/llvm-project/blob/main/libc/src/__support/FPUtil/FPBits.h#L736

Sh0g0-1758 replied:
done.

lntue commented on libc/src/__support/FPUtil/BasicOperations.h:
```diff
@@ -73,6 +76,54 @@ LIBC_INLINE T fdim(T x, T y) {
+        cx = FPBits<T>::signaling_nan().get_val();
+      }
+    } else if (exponent == 0 && bit63) {
+      cx = FPBits<T>::get_canonical_val(sx.sign(), mantissa).get_val();
```
Use `FPBits<T>::make_value` instead: https://github.com/llvm/llvm-project/blob/main/libc/src/__support/FPUtil/FPBits.h#L736

Sh0g0-1758 replied:
done.

lntue commented on libc/src/__support/FPUtil/FPBits.h:
```diff
@@ -501,6 +501,10 @@ struct FPRepSem<FPType::X86_Binary80, RetT>
                        Significand::msb() | (Significand::msb() >> 1) |
                            Significand(v)));
   }
+  LIBC_INLINE static constexpr RetT get_canonical_val(Sign sign = Sign::POS,
+                                                      StorageType v = 0) {
+    return RetT(encode(sign, Exponent::min(), Significand(v)));
+  }
```
Remove this function, use `make_value` instead.

Sh0g0-1758 replied:
done.

lntue commented on libc/src/__support/FPUtil/BasicOperations.h:
```diff
@@ -73,6 +76,54 @@ LIBC_INLINE T fdim(T x, T y) {
+    int exponent = sx.get_biased_exponent();
+    if (exponent == 0x7FFF) {
+      if (!bit63 && !bit62) {
+        if (!bit61) {
```
for single statement after `if-else`, you don't need curly braces.

Sh0g0-1758 replied:
done.

lntue commented on libc/test/src/math/smoke/CanonicalizeTest.h:
```diff
@@ -0,0 +1,54 @@
+  void testSpecialNumbers(CanonicalizeFunc func) {
+    EXPECT_FP_EQ(zero, func(zero));
+    EXPECT_FP_EQ(neg_zero, func(neg_zero));
+
```
Add test for signaling nan.

Sh0g0-1758 replied:
done.

lntue commented:
Can you also update the status page at `libc/docs/math/index.rst`? Thanks,

Sh0g0-1758 committed:
minor changes

Sh0g0-1758 committed:
minor bug fix

Sh0g0-1758 committed:
Added tests

Sh0g0-1758 committed:
minor bug fix

Sh0g0-1758 committed:
scope fix

Sh0g0-1758 committed:
minor fix

Sh0g0-1758 committed:
Added spec

Sh0g0-1758 committed:
Formatted docs

Sh0g0-1758 commented:
@Intue, is there something I can use for formatting rst? It seems a lot of manual labor is involved to refactor the table.

Sh0g0-1758 committed:
restructure table

lntue commented:
> @lntue , is there something I can use for formatting rst? It seems a lot of manual labor is involved to refactor the table.
> 
> Nvm: DId it anyways :)

Unfortunately, right now copy-paste is the best way to edit rst that I know.

Sh0g0-1758 commented:
@Intue, are any further changes required?

lntue commented on libc/src/__support/FPUtil/BasicOperations.h:
```diff
@@ -73,6 +76,56 @@ LIBC_INLINE T fdim(T x, T y) {
+LIBC_INLINE int canonicalize(T &cx, const T &x) {
+  FPBits<T> sx(x);
+  if constexpr (get_fp_type<T>() == FPType::X86_Binary80) {
+    // All the pseudo and unnormal numbers are not canonical.
```
Can you add tests for all these special cases for x86-long double?  Those tests should be guarded by `LIBC_TYPES_LONG_DOUBLE_IS_X86_FLOAT80`.

Sh0g0-1758 replied:
I am not sure how to add those test cases, can you give me an example for this : 

```
    //            |   Bit 63   | Bits 62-0 |
    // All zeroes |   One      | Anything  | Pseudo Denormal, Value =
    //            |            |           | (−1)**s × m × 2**−16382
```

lntue replied:
You can add some testing method in `CanonicalizeTest` class that is only available for x86 long double type, in which you can build these non-canonical bit patterns with `UInt128(exponent) << 64 + UInt128(bits)`, feed them into `FPBits`, and extract the long double value to be called by `canonicalizel`.

Then in `canonicalizel_test.cpp`, you just add an extra `TEST_F` calling that test method, that is wrapped under `#ifdef LIBC_TYPES_LONG_DOUBLE_IS_X86_FLOAT80`.

Sh0g0-1758 replied:
Thanks for the clarification. Added tests for the same.

lntue commented on libc/docs/math/index.rst:
```diff
@@ -104,241 +104,249 @@ Implementation Status
-+--------------+---------+---------+---------+---------+---------+---------+---------+---------+---------+---------+---------+---------+
-| truncf128    | |check| | |check| |         | |check| |         |         |         |         |         |         |         |         |
-+--------------+---------+---------+---------+---------+---------+---------+---------+---------+---------+---------+---------+---------+
++------------------+---------------------------------------+-------------------+-------------------+-------------------+-------------------+
```
Can you test the generated docs as follow:
- add `-DLLVM_ENABLE_SPHINX=ON -DLIBC_INCLUDE_DOCS=ON` to your cmake command
- run `ninja docs-libc-html`
- check the generated page.

Sh0g0-1758 replied:
It is building fine.

Sh0g0-1758 committed:
Added more tests

lntue commented on libc/test/src/math/smoke/CanonicalizeTest.h:
```diff
@@ -45,6 +48,104 @@ class CanonicalizeTest : public LIBC_NAMESPACE::testing::Test {
     EXPECT_EQ(cx, -aNaN);
   }
 
+  void testX64_80SpecialNumbers(CanonicalizeFunc f) {
```
Put the whole test under: `if constexpr( LIBC_NAMESPACE::fputil::get_fp_type() == ...) { ... }`

Sh0g0-1758 replied:
done.

Sh0g0-1758 committed:
updated tests

lntue commented on libc/test/src/math/smoke/CanonicalizeTest.h:
```diff
@@ -0,0 +1,192 @@
+  }
+
+  if constexpr (LIBC_NAMESPACE::fputil::get_fp_type() == FPType::X86_Binary80) {
+    void testX64_80SpecialNumbers(CanonicalizeFunc f) {
```
the `if constexpr` should be inside the test function body.

Sh0g0-1758 replied:
Ah right, put the ordering wrong.

Sh0g0-1758 committed:
Update CanonicalizeTest.h

Sh0g0-1758 committed:
Ran formatter

Sh0g0-1758 commented:
Any other major changes?

lntue commented on libc/src/__support/FPUtil/CMakeLists.txt:
```diff
@@ -82,8 +82,11 @@ add_header_library(
     libc.src.__support.CPP.type_traits
+    libc.src.__support.uint128
     libc.src.__support.common
+    libc.src.__support.macros.optimization.h
```
I was curious why your tests were not running in the buildkite pre-submit check.  Turn out I missed this one:
it should be `libc.src.__support.macros.optimization`, without `.h`.  After updating this dependency, you should be able to build and see some of the issues with the tests when running `ninja libc`.

Sh0g0-1758 replied:
Ah so that was it. I was wondering the same thing since morning when I took a look at the buildkite pre-submit check. Thanks for pointing it out.

Sh0g0-1758 committed:
bug fix

Bug fixes

Ran formatter

Updated tests

Ran formatter

bug fix

Bug fixes

Minor bug fixes

Bug Fixes

Add namespace

Bug Fix

Changed namespace

Minor changes

Experimenting

revert

changed type

Update tests

Added types

Update BasicOperations.h

Bug Fix

Formatter

minor changes

pushing all tests

Formatting

revert

Bug fix

Minor changes

changes

Formatter

Testing

changes

Added a test

Ran formatter

Minor push

formatter

fix

Typo

Removed bit 61

revamp testing

TEst 4

changes

added type

Formatter

Test

Bug fix

Bug fix

Fixed bug

Formatter

nits

nits

info

RE

s

s

s

s

TEst

TEst

Added test 5

F

minor changes

nits

nits

bit62 fix

testing

revert

nots

fix

nits

Format

fix

F

test

test

Buf Fix

Sh0g0-1758 commented:
@Intue, I have fixed most of the tests. The only cases remaining are test_1 and test_5 for X86_80SpecialTests where for some reason the matching is not happening correctly and aNaN is getting returned. Can you take a look at those tests once, I can't seem to figure out the reason.

Sh0g0-1758 commented:
Ah I got what was causing the error in the tests. Fixed them.

Sh0g0-1758 committed:
nits

lntue commented on libc/src/__support/FPUtil/BasicOperations.h:
```diff
@@ -9,10 +9,13 @@
 #ifndef LLVM_LIBC_SRC___SUPPORT_FPUTIL_BASICOPERATIONS_H
 #define LLVM_LIBC_SRC___SUPPORT_FPUTIL_BASICOPERATIONS_H
 
+#include "FEnvImpl.h"
 #include "FPBits.h"
 
 #include "src/__support/CPP/type_traits.h"
+#include "src/__support/UInt128.h"
 #include "src/__support/common.h"
+#include "src/__support/macros/optimization.h" // LIBC_UNLIKELY
```
Thanks for fixing all the issues!  Everything is built and tested fine on my machine now.

One last thing, do you mind adding:
```
        ":__support_fputil_fenv_impl",
        ":__support_macros_optimization",
        ":__support_uint128",
```
to the dependency of `__support_fputil_basic_operations` in the project's bazel overlay at:
https://github.com/llvm/llvm-project/blob/main/utils/bazel/llvm-project-overlay/libc/BUILD.bazel#L624

and sync to head to resolve the merge conflicts.  After that the patch should be ready to be merged.

Thanks,

Sh0g0-1758 replied:
done.

Sh0g0-1758 committed:
Update bazel

Sh0g0-1758 committed:
Merge branch 'main' into 85286

lntue reviewed (approved):
Thanks for the patch!

nickdesaulniers commented:
This breaks the arm32 build in postsubmit, PTAL.
https://lab.llvm.org/buildbot/#/builders/229/builds/24394
```
In file included from /llvm/libc_worker/worker/libc-arm32-debian/libc-arm32-debian-dbg/llvm-project/libc/src/math/generic/sinf.cpp:11:
/llvm/libc_worker/worker/libc-arm32-debian/libc-arm32-debian-dbg/llvm-project/libc/src/__support/FPUtil/BasicOperations.h:202:10: error: no viable conversion from '__llvm_libc_19_0_0_git::BigInt<128, false>' to 'bool'
    bool bit62 = ((mantissa & (1ULL << 62)) >> 62);
         ^       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/llvm/libc_worker/worker/libc-arm32-debian/libc-arm32-debian-dbg/llvm-project/libc/src/__support/UInt.h:186:56: note: explicit conversion function is not a candidate
  template <typename T> LIBC_INLINE constexpr explicit operator T() const {
                                                       ^
/llvm/libc_worker/worker/libc-arm32-debian/libc-arm32-debian-dbg/llvm-project/libc/src/__support/UInt.h:216:34: note: explicit conversion function is not a candidate
  LIBC_INLINE constexpr explicit operator bool() const { return !is_zero(); }
                                 ^
```

nickdesaulniers commented:
I wonder why UInt::operator bool is marked explicit? Implicit conversions to bool sure are nice.

Sh0g0-1758 commented:
I can provide a patch but I don't have context. Can start working on it once you provide some pointers.

michaelrj-google commented:
Since the issue is that the implicit conversion to bool is failing, the easiest solution is likely to add a `static_cast<bool>` to the statement.

nickdesaulniers commented:
> Since the issue is that the implicit conversion to bool is failing, the easiest solution is likely to add a `static_cast<bool>` to the statement.

Cant leave the buildbots red for long: https://github.com/llvm/llvm-project/pull/86736

Sh0g0-1758 commented:
Ah sorry for the delay. My OS crashed and I saw the dreaded white death screen. Had to set up my system from scratch. Thanks for the patch @nickdesaulniers.

jcranmer-intel commented on libc/src/__support/FPUtil/BasicOperations.h:
```diff
@@ -178,6 +181,66 @@ LIBC_INLINE T fdim(T x, T y) {
+    // More precisely :
+    // Exponent   |       Significand      | Meaning
+    //            | Bits 63-62 | Bits 61-0 |
+    // All Ones   |     00     |    Zero   | Pseudo Infinity, Value = Infinty
```
This is not correct.

The x86 manual states:
> Specifically, the categories of encodings formerly known as pseudo-NaNs, pseudo-infinities, and un-normal
numbers are not supported and should not be used as operand values. The Intel 387 math coprocessor and later
IA-32 processors generate an invalid-operation exception when these encodings are encountered as operands.

So pseudo-infinity and pseudo-denormal should be canonicalized to sNaN, not infinity or finite value (respectively).

Sh0g0-1758 replied:
Ah I see, from the wikipedia page : 

> Pseudo-Infinity. The sign bit gives the sign of the infinity. The 8087 and 80287 treat this as Infinity. The 80387 and later treat this as an invalid operand.

I focused only the 8087 and 80287 part. Will link a patch soon.

lntue replied:
@jcranmer-intel : do you mean to treat them as `sNaN`, not literally return `sNaN`?  B/c `canonicalize` translates `sNaN` to `qNaN + FE_INVALID`.

Sh0g0-1758 replied:
@jcranmer-intel, can you confirm the above.

jcranmer-intel replied:
Yes, they should be treated as sNaN, and therefore returned as qNaN

(alternatively, use the escape hatch of don't-canonicalize-this-value that C provides--glibc takes this approach).
